File: problems/HackerRank/countingSort.ts
function countingSort(arr: number[]): number[] {
  // [1,1,3,2,1]

  // [0,1,0,0]
  // [0,2,0,0]
  // [0,2,0,1]
  // [0,2,1,1]
  // [0,3,1,1]

  // [1,1,1,2,3]

  const array = Array(10).fill(0);

  for (let i = 0; i < arr.length; i++) {
    const ele = arr[i];

    if (array[ele] !== 0) {
      array[ele]++;
      continue;
    }
    array[ele] = 1;
  }

  return array;

  const result = [] as number[];

  for (let i = 0; i < array.length; i++) {
    let ele = array[i];
    while (ele > 0) {
      result.push(i);
      ele--;
    }
  }

  return result;
}

console.log(countingSort([1, 1, 3, 2, 1]));

File: problems/HackerRank/week2/countingSort.py
def countingSort(arr):
    ans = [0] * 100
    
    for num in arr:
        ans[num] += 1
    
    return ans
    
    
# Complete version
def countingSort(arr):
    # Initialize frequency array for numbers 0 to 99
    count = [0] * 100
    
    # Count the occurrences of each number in arr
    for num in arr:
        count[num] += 1
    
    # Reconstruct the sorted array from the frequency counts
    sorted_arr = []
    for i in range(len(count)):
        sorted_arr.extend([i] * count[i])
    
    return sorted_arr
